### PR TITLE
Fused QKV GQAttention implementation

### PR DESCRIPTION
**Pull request by wwwjn**
## Implementation
Replaces separate wq, wk, wv with single wqkv linear of shape [n_kv_heads * R * head_dim, dim], where R = heads_per_kv + 2. Fused q/k/v project to reduce number of matmul kernel launches, and also reduce the number of TP communication kernel launches.

Model-wise: 
- DeepSeek-v3 is untouched as the Attention has 2 different computation path (depend on `q_lora_rank`), makes it not easy to interact with fusing decision. 
- Flux is also untouched
 

## Test 
using llama3 debug model:
- Numerics test: When initialized using the same seed checkpoint: 
Screenshot 2026-04-07 at 6 02 38 PM

- Performance test fused v.s. unfused: 
    - CPU bounded model: Screenshot 2026-04-07 at 6 04 39 PM
    - GPU (memory-bandwidth bounded, or GPU compute bounded): Screenshot 2026-04-07 at 7 22 47 PM

- Test with DCP + state_dict_adapter: 
Screenshot 2026-04-09 at 2 19 35 PM

### CODE REVIEW

File: torchtitan/models/common/attention.py [hunk 1]
```diff
@@ -622,12 +619,15 @@ def __init__(self, config: Config):
         # Scaling factor (needed when head_dim differs from dim // n_heads)
         self.scaling = self.head_dim**-0.5 if config.head_dim is not None else None
 
-        self.wq = config.wq.build()
-        self.wk = config.wkv.build()
-        self.wv = config.wkv.build()
-        self.wo = config.wo.build()
+    def _project_qkv(
```
Comment (tianyu-l) [inline torchtitan/models/common/attention.py]: I think here composition makes more sense than inheritance, because essentially we are just swapping the qkv projection, but keeping all other implementation untouched.

Can we introduce a single `wqkv` in the config, of class `QKVLinear`, which can have a subclass `FusedQKVLinear`?

File: torchtitan/models/common/attention.py [hunk 2]
```diff
@@ -675,4 +669,72 @@ def forward(
...
+        return self.wo(output)  # pyrefly: ignore [not-callable]
+
+
+class GQAttention(BaseGQAttention):
+    """Grouped-Query Attention with separate Q, K, V projections."""
+
+    @dataclass(kw_only=True, slots=True)
+    class Config(BaseGQAttention.Config):
+        wq: Linear.Config
+        wkv: Linear.Config
```
Comment (tianyu-l) [inline torchtitan/models/common/attention.py]: The reason of keeping them separately might be that they have different `param_init`? I think we should remove depth_init and thus may be able to make them into one `wqkv`.

See discussions in https://github.com/pytorch/torchtitan/issues/375#issuecomment-4187484549

It will change numerics, but it's fine.

File: torchtitan/models/gpt_oss/state_dict_adapter.py
Comment (tianyu-l) [inline torchtitan/models/gpt_oss/state_dict_adapter.py:1]: are these changes tested with TP applied on DCP side?

File: torchtitan/models/llama3/parallelize.py [hunk 3]
```diff
@@ -217,18 +217,33 @@ def apply_tp(
         output_layouts=sp_layout, use_local_output=enable_sp
     )
 
+    # Detect whether fused QKV is used by checking the first layer
+    # pyrefly: ignore [not-callable]
+    first_block = next(iter(model.layers.values()))
+    use_fused_qkv = hasattr(
+        first_block.attention, "wqkv"  # pyrefly: ignore [missing-attribute]
```
Comment (tianyu-l) [inline torchtitan/models/llama3/parallelize.py]: use `isinstance` not `hasattr`

File: torchtitan/models/common/attention.py (hunk 1 above)
Comment (wwwjn) [reply torchtitan/models/common/attention.py]: I see, the refined design would be:

- Only one attention GQAttention class, has a field `wqkv: BaseQKVLinear`
- Class `QKVLinear` is a subclass of `BaseQKVLinear`,  and has fields `wq` / `wk` / `wv`
- Class `FusedQKVLinear` is a subclass of `BaseQKVLinear`,  and has fields `wqkv`

File: torchtitan/models/common/attention.py (hunk 2 above)
Comment (wwwjn) [reply torchtitan/models/common/attention.py]: The reason is only about `depth_init`, also becuase the wk and wv has the same number of heads (num_kv_heads), but in GQAttention, `num_q_heads` not necessarily equal to num_kv_heads. So wq has different Linear.out_features

File: torchtitan/models/common/attention.py [hunk 4]
```diff
@@ -565,23 +568,106 @@ def __post_init__(self):
...
+    :class:`FusedQKVLinear` for a single fused projection.
     """
 
     @dataclass(kw_only=True, slots=True)
     class Config(BaseAttention.Config):
         n_heads: int
         dim: int
-        wq: Linear.Config
-        wkv: Linear.Config
+        qkv: BaseQKVLinear.Config
```
Comment (tianyu-l) [inline torchtitan/models/common/attention.py]: qkv sounds ambiguous, how about
```suggestion
        qkv_linear: BaseQKVLinear.Config
```

File: torchtitan/models/common/attention.py [hunk 5]
```diff
@@ -565,23 +568,106 @@ def __post_init__(self):
...
+        wkv: Linear.Config
+
+    def __init__(self, config: Config):
+        super().__init__()
+        self.wq = config.wq.build()
+        self.wk = config.wkv.build()
+        self.wv = config.wkv.build()
+
+    def forward(
+        self, x: torch.Tensor, bs: int, seqlen: int, head_dim: int
```
Comment (tianyu-l) [inline torchtitan/models/common/attention.py]: Let's not pass bs / seqlen / head_dim in forward. Instead
- put `head_dim` in config
- infer bs and seqlen from x shape

File: torchtitan/models/common/attention.py [hunk 6]
```diff
@@ -565,23 +568,106 @@ def __post_init__(self):
...
+
+    def __init__(self, config: Config):
+        super().__init__()
+        self.wqkv = config.wqkv.build()
+        self.heads_per_kv = config.heads_per_kv
+        # R = Q-heads-per-KV-group + 1 (K) + 1 (V)
+        self.r_dim = config.heads_per_kv + 2
+
+    def forward(
+        self, x: torch.Tensor, bs: int, seqlen: int, head_dim: int
```
Comment (tianyu-l) [inline torchtitan/models/common/attention.py]: similarly, I think it's better to store `head_dim` (if we put it in Base class config, we automatically get it here).

File: torchtitan/models/qwen3/__init__.py [hunk 7]
```diff
@@ -456,6 +458,44 @@ def _8b_varlen() -> Qwen3Model.Config:
     return config
 
 
+def _4b_fused_qkv() -> Qwen3Model.Config:
```
Comment (tianyu-l) [inline torchtitan/models/qwen3/__init__.py]: let's create a debug model with fused_qkv and swap one qwen3 model CI test (maybe the moe one) to guard on it

Comment (wwwjn) [reply torchtitan/models/qwen3/__init__.py]: Nice catch, this config is for perf testing and I removed it.

File: tests/integration_tests/models.py [hunk 8]
```diff
@@ -147,6 +147,18 @@ def build_model_tests_list() -> list[OverrideDefinitions]:
             "qwen3_fsdp+tp+ep+etp",
             ngpu=4,
         ),
+        OverrideDefinitions(
+            [
+                [
+                    "--module qwen3 --config qwen3_debugmodel_fused_qkv",
```
Comment (tianyu-l) [inline tests/integration_tests/models.py]: Can we piggyback on existing tests instead of creating a new one? I'm worried about CI pressure in general.
